core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -274,62 +327,133 @@ sealed trait Cause[+E] extends Product with Serializable { self =>
+      case Cause.Fail(None)           => None
+      case Cause.Then(left, right) =>
+        (sequenceCauseOption(left), sequenceCauseOption(right)) match {
+          case (Some(cl), Some(cr)) => Some(Cause.Then(cl, cr))
+          case (None, Some(cr))     => Some(cr)
+          case (Some(cl), None)     => Some(cl)
+          case (None, None)         => None
+        }
+
+      case Cause.Both(left, right) =>
+        (sequenceCauseOption(left), sequenceCauseOption(right)) match {
+          case (Some(cl), Some(cr)) => Some(Cause.Both(cl, cr))
+          case (None, Some(cr))     => Some(cr)
+          case (Some(cl), None)     => Some(cl)
+          case (None, None)         => None
+        }
+    }
 
-  final case class Fail[E](value: E) extends Cause[E] {
+  private case class Fail[E](value: E) extends Cause[E] {
     override final def equals(that: Any): Boolean = that match {
       case fail: Fail[_]     => value == fail.value
       case traced: Traced[_] => this == traced.cause
+      case meta: Meta[_]     => this == meta.cause
       case _                 => false
     }
   }
 
-  final case class Die(value: Throwable) extends Cause[Nothing] {
+  object Fail {
+    def apply[E](value: E): Cause[E] =
+      new Fail(value)
+  }
+
+  private case class Die(value: Throwable) extends Cause[Nothing] {
     override final def equals(that: Any): Boolean = that match {
       case die: Die          => value == die.value
       case traced: Traced[_] => this == traced.cause
+      case meta: Meta[_]     => this == meta.cause
       case _                 => false
     }
   }
 
+  object Die {
+    final def apply(value: Throwable): Cause[Nothing] =
+      new Die(value)
+  }
+
   case object Interrupt extends Cause[Nothing] {
     override final def equals(that: Any): Boolean =
       (this eq that.asInstanceOf[AnyRef]) || (that match {
         case traced: Traced[_] => this == traced.cause
+        case meta: Meta[_]     => this == meta.cause
         case _                 => false
       })
   }
 
   // Traced is excluded completely from equals & hashCode
-  final case class Traced[E](cause: Cause[E], trace: ZTrace) extends Cause[E] {
+  private case class Traced[E](cause: Cause[E], trace: ZTrace) extends Cause[E] {
     override final def hashCode: Int = cause.hashCode()
     override final def equals(obj: Any): Boolean = obj match {
       case traced: Traced[_] => cause == traced.cause
+      case meta: Meta[_]     => cause == meta.cause
+      case _                 => cause == obj
+    }
+  }
+
+  object Traced {
+    def apply[E](cause: Cause[E], trace: ZTrace): Cause[E] =
+      new Traced(cause, trace)
+  }
+
+  // Meta is excluded completely from equals & hashCode
+  private case class Meta[E](cause: Cause[E], data: Data) extends Cause[E] {
+    override final def hashCode: Int = cause.hashCode
+    override final def equals(obj: Any): Boolean = obj match {
+      case traced: Traced[_] => cause == traced
```

Comment:
```suggestion
      case traced: Traced[_] => cause == traced.cause
```

Reply:
Traced could probably be replaced by Meta, tbh...

Reply:
Yes.